scipy/stats/_stats_py.py

```diff
@@ -3908,34 +3922,80 @@ def f_oneway(*samples, axis=0):
 
-    normalized_ss = _square_of_sums(alldata, axis=axis) / bign
+    if equal_var:
+        # Determine the mean of the data, and subtract that from all inputs to a
+        # variance (via sum_of_sq / sq_of_sum) calculation.  Variance is invariant
+        # to a shift in location, and centering all data around zero vastly
+        # improves numerical stability.
+        offset = alldata.mean(axis=axis, keepdims=True)
+        alldata = alldata - offset
+
+        normalized_ss = _square_of_sums(alldata, axis=axis) / bign
+
+        sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+
+        ssbn = 0
+        for sample in samples:
+            smo_ss = _square_of_sums(sample - offset, axis=axis)
+            ssbn = ssbn + smo_ss / sample.shape[axis]
+
+        # Naming: variables ending in bn/b are for "between treatments", wn/w are
+        # for "within treatments"
+        ssbn = ssbn - normalized_ss
+        sswn = sstot - ssbn
+        dfbn = num_groups - 1
+        dfwn = bign - num_groups
+        msb = ssbn / dfbn
+        msw = sswn / dfwn
+        with np.errstate(divide='ignore', invalid='ignore'):
+            f = msb / msw
 
-    sstot = _sum_of_squares(alldata, axis=axis) - normalized_ss
+        prob = special.fdtrc(dfbn, dfwn, f)   # equivalent to stats.f.sf
 
-    ssbn = 0
-    for sample in samples:
-        smo_ss = _square_of_sums(sample - offset, axis=axis)
-        ssbn = ssbn + smo_ss / sample.shape[axis]
-
-    # Naming: variables ending in bn/b are for "between treatments", wn/w are
-    # for "within treatments"
-    ssbn = ssbn - normalized_ss
-    sswn = sstot - ssbn
-    dfbn = num_groups - 1
-    dfwn = bign - num_groups
-    msb = ssbn / dfbn
-    msw = sswn / dfwn
-    with np.errstate(divide='ignore', invalid='ignore'):
-        f = msb / msw
+    # if population variances are not equeal, perform Welch ANOVA
+    # the related issue: https://github.com/scipy/scipy/issues/11122
+    else:
+        # calculate basic statistics for each sample
+        means = np.asarray([np.mean(sample, axis=axis) for sample in samples])
+        ns = np.asarray(
+            [np.apply_along_axis(len, axis=axis, arr=sample) for sample in samples])
+        k = len(samples)
+
+        vars_ = np.asarray([np.var(sample, axis=axis, ddof=1) for sample in samples])
+
+        # calculate weight by number of data and varianc
+        # ref.[4] p.330 ($$w = 1/{\lambda s^2}$$ where $$\lambda=1/n$$ in ref)
+        ws = ns / vars_
+
+        # calculate adjusted grand mean
+        # ref.[4] p.330 (notation is $$\hat y$$ in ref)
+        mean_prime = np.sum(ws * means, axis=0) / np.sum(ws, axis=0)
+
+        # adjust f statistic
+        # ref.[4] p.334 eq.29
+        numerator = np.sum(ws * (means - mean_prime)**2, axis=0) / (k - 1)
+        denominator = (
+                1 + 2 * (k - 2) / (k**2 - 1) *
+                np.sum((1 / (ns - 1)) *
+                       (1 - ws / np.sum(ws, axis=0))**2,
```

Comment:
```suggestion
                       (1 - ws / S_ws)**2,
```
Let's create a variable `S_ws = np.sum(ws, axis=0)` and use it here and below. I think it will make the equations ever so slightly easier to compare against the reference, and it's a little faster.